### PR TITLE
test: added few taproot integration test case

### PR DESCRIPTION
**Pull request by 0xEgao**
This PR works on adding the remaining edge cases to test to ensure the Taproot-based Coinswap protocol works effectively and safely under both normal and failure conditions.

Works on issue #658

Already Covered Test Cases
---
* [x] Two maker single taker swap
* [x] Maker abort after `SenderContract` from Taker
* [x] Intermediate maker abort after `SenderContractFromMaker`
* [x] Final maker abort after `SenderContractFromMaker`
* [x] Incoming contract timelock broadcast before `PrivateKeyHandover`
* [x] Outgoing contract hashlock broadcast before `PrivateKeyHandover`

Added in This PR
---

* [x] Multi maker single taker swap
* [x] Multi taker multi maker swap
* [x] Not enough makers
* [x] Maker abort after `AckResponse` on `SwapDetails`
* [x] Maker abort after `PrivateKeyHandover` by Taker
* [x] Taker abort after `AckResponse` from Maker
* [x] Taker abort after `SenderContract` from Taker
* [x] Taker abort after `SenderContractFromMaker`

### CODE REVIEW

**Review by mojoX911 — CHANGES_REQUESTED** [review]
Overall Ack. Just one major comment for all the test cases.

Instead of asserting for a range (< or >) can we assert for exact values? For all swaps that failed, succeeded, or recovered, the difference between balances should always be a fixed constant number. So we can assert for that number directly, and also mention in a comment why this difference in balance is occurring. 

Currently, we are not able to see how much exactly the takers and makers are losing in each scenario.

File: tests/test_framework/mod.rs [hunk 1]
```diff
@@ -560,6 +625,55 @@ pub fn verify_maker_pre_swap_balances(balances: &Balances, assert_regular_balanc
...
+            i, balances.regular, balances.swap, balances.contract, balances.fidelity
+        );
+
+        // Regular balance after fidelity bond (allow small fee variance)
+        // eg for 0.20 btc as funded, 0.05 btc will be spent due to fidelity bond creation and some small fee,so regular balance will be around 0.1499
+        assert!(
+            balances.regular.to_sat() == 14999510    // 1st maker  (normal case)
+            || balances.regular.to_sat() == 14999508 // 2nd maker  (normal case)
+            || balances.regular.to_sat() == 34999510 // 1st maker multi taker case (8 utxo funded)
+            || balances.regular.to_sat() == 34999508, // 2nd maker multi taker case (8 utxo funded)
```
Comment (mojoX911) [inline tests/test_framework/mod.rs]: These are the 2sats difference for p2wsh signature grinding. So it can be covered by the `assert_in_range` macro.

Our fidelity bonds are still p2wsh even for taproot makers, thats why this variance is coming.  For now its okay. 

A worthy thing to fix post-release.

File: tests/taproot_swap.rs [hunk 2]
```diff
@@ -305,25 +176,32 @@ fn verify_taproot_swap_results(
...
-        assert!(
-            balances.spendable >= expected_minimum,
-            "Taproot Maker {} balance should not decrease significantly. Original spendable: {}, Current spendable: {}, Expected minimum: {}",
-            i,
-            original_spendable,
-            balances.spendable,
-            expected_minimum
+        assert_in_range!(
+            balances.spendable.to_sat(),
+            [15020199, 15031704], // here no fund loss because swap never happen
```
Comment (mojoX911) [inline tests/taproot_swap.rs]: "swap never happens"? Balances did increase. The comment should be something like "makers have more than 15 million sats as they earned swap fees".

File: tests/taproot_swap.rs [hunk 3]
```diff
@@ -171,131 +128,45 @@ fn test_taproot_coinswap() {
...
-
-    assert_eq!(balances.regular, expected_regular);
-
+    let taker_balances_after = taproot_taker.get_wallet().get_balances().unwrap();
+    info!("📊 Taproot Taker balance after completing swap:");
     info!(
-        "Taproot Taker funded successfully. Regular: {}, Spendable: {}",
-        balances.regular, balances.spendable
+        "  Regular: {}, Contract: {}, Spendable: {}",
+        taker_balances_after.regular, taker_balances_after.contract, taker_balances_after.spendable
```
Comment (mojoX911) [inline tests/taproot_swap.rs]: We should add the swap balance too in the display.

File: tests/test_framework/mod.rs (hunk 1 above)
Comment (0xEgao) [reply tests/test_framework/mod.rs]: Done, have replaced this with `assert_in_range`.

File: tests/taproot_swap.rs (hunk 3 above)
Comment (mojoX911) [reply tests/taproot_swap.rs]: Samer for other tests.

File: tests/taproot_swap.rs [hunk 4]
```diff
@@ -80,45 +75,7 @@ fn test_taproot_coinswap() {
...
-        assert!(
-            balances.spendable > Amount::ZERO,
-            "Maker should have spendable balance"
-        );
-
-        // Store the actual spendable balance AFTER fidelity bond creation
-        actual_maker_spendable_balances.push(balances.spendable);
-    }
-
+    let actual_maker_spendable_balances = verify_maker_pre_swap_balance_taproot(&taproot_makers);
```
Comment (mojoX911) [inline tests/taproot_swap.rs]: nit: why `actual`? Just call it `maker_spendable_balance`.

File: tests/taproot_maker_abort1.rs [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+        let balance_diff = balances
+            .spendable
+            .to_sat()
+            .saturating_sub(original_spendable.to_sat());
+        // maker gained fee
+        assert_in_range!(
+            balance_diff,
+            [0],
+            "Taproot Maker shouldn't have gain any fee here"
```
Comment (mojoX911) [inline tests/taproot_maker_abort1.rs]: Assert in range not required here as it will always be 0.

File: tests/taproot_maker_abort2.rs [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+    assert!(
+        taker_total_after.to_sat() == 14944003,
+        "Taproot Taker Balance check after hashlock recovery. Original: {}, After: {}",
+        taproot_taker_original_balance,
+        taker_total_after
+    );
+
+    let balance_diff = taproot_taker_original_balance - taker_total_after;
+    assert!(
+        balance_diff.to_sat() == 55997, // Hashlock recovery fee
```
Comment (mojoX911) [inline tests/taproot_maker_abort2.rs]: This is not really just the hashlock fee. This is maker_fee + hashlock_tx_fee. Would be good to separate these two out. Just assert it as sum of those two numbers and explain in the comment.

Comment (mojoX911) [reply tests/taproot_maker_abort2.rs]: Similar for other tests. Break down the total fee components.

File: tests/taproot_maker_abort2.rs [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+    info!(
+        "  Regular: {}, Contract: {}, Spendable: {}",
+        taker_balances_after.regular, taker_balances_after.contract, taker_balances_after.spendable
+    );
+
+    // Now wait for maker to extract preimage and recover via hashlock
+    info!("⏳ Waiting for maker to extract preimage and recover via hashlock...");
+    std::thread::sleep(std::time::Duration::from_secs(60));
+    // Mine more blocks to give maker time to see the hashlock sweep
+    generate_blocks(bitcoind, 2);
```
Comment (mojoX911) [inline tests/taproot_maker_abort2.rs:130]: Instead of sleeping, we should actually watch for the right log. We already have `assert_log()` for the test framework. Made exactly for these behavior assertions. 

Applies to all other places where tests are sleeping for something to get done that can be identified in the logs.

**Review by mojoX911 — CHANGES_REQUESTED** [review]

File: tests/taproot_maker_abort2.rs [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    let taker_balances_after = taproot_taker.get_wallet().get_balances().unwrap();
+    info!(
+        "  Regular: {}, Contract: {}, Spendable: {}",
+        taker_balances_after.regular, taker_balances_after.contract, taker_balances_after.spendable
+    );
+
+    // Now wait for maker to extract preimage and recover via hashlock
+    info!("⏳ Waiting for maker to extract preimage and recover via hashlock...");
+    std::thread::sleep(std::time::Duration::from_secs(60));
+    // Mine more blocks to give maker time to see the hashlock sweep
```
Comment (mojoX911) [inline tests/taproot_maker_abort2.rs:129]: One more behavior to check for at this stage. Start the faulty maker again. It should trigger its recovery routine, get the hash preimage from the blockchain, and sweep the funds.

We need to assert that too. If that doesn't happen there could be fund theft by the taker/or the other maker, sweeping back the contract via timelock.

Comment (mojoX911) [reply tests/taproot_maker_abort2.rs:129]: Applies to other tests too. Start back the faullty maker and assert it behaved as expected.

File: tests/taproot_maker_abort2.rs [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+        );
+
+        // Use spendable (regular + swap) for comparison
+        // Here in hashlock recovery the spendable balance is almost similar to key-path spend
+        // as the parties are completing their swap by claiming their incoming contract
+        // via script-path spend.
+        assert_in_range!(
+            balances.spendable.to_sat(),
+            [14999510, 15020999, 15032506],
+            "Taproot Maker after hashlock recovery balance check."
```
Comment (mojoX911) [inline tests/taproot_maker_abort2.rs:189]: Why 3 balances here?

File: tests/taproot_swap.rs [hunk 10]
```diff
@@ -305,25 +177,32 @@ fn verify_taproot_swap_results(
...
-        // The maker's spendable balance should be at least (original - max_fees + min_earned_fees)
-        let expected_minimum = original_spendable - max_fees;
-        assert!(
-            balances.spendable >= expected_minimum,
-            "Taproot Maker {} balance should not decrease significantly. Original spendable: {}, Current spendable: {}, Expected minimum: {}",
-            i,
-            original_spendable,
-            balances.spendable,
-            expected_minimum
+        assert_in_range!(
```
Comment (mojoX911) [inline tests/taproot_swap.rs:180]: I think this can also be just `assert` for all tests.

File: tests/taproot_swap.rs
Comment (mojoX911) [inline tests/taproot_swap.rs:106]: This is a good spot to assert that swaped failed with the expected error message. 

Doesn't apply for these tests, but applies for all other cases.

File: tests/taproot_maker_abort2.rs [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+    };
+
+    // Attempt the swap - it will fail when maker closes connection
+    // After recovery, do_coinswap returns Ok(None) to indicate recovery was triggered
+    match taproot_taker.do_coinswap(swap_params) {
+        Ok(Some(_report)) => {
+            panic!("Swap should have failed due to maker closing connection, but succeeded with report!");
+        }
+        Ok(None) => {
+            info!("✅ Taproot coinswap triggered recovery as expected (Ok(None))");
```
Comment (mojoX911) [inline tests/taproot_maker_abort2.rs:106]: This test actually returns None, while we should consider it as a valid swap too. It just has script_path spend instead of key_path.

This can be fixed later. Noting it down so we can open an Issue and track. `Fix: Return swap report even if recovered via hashlock`.

File: tests/taproot_multi_maker.rs [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+        );
+
+        // In multi maker it's hard to track the absolute spendable balance of a
+        // particular maker, as each time their order can be different leading to
+        // differnt spendable balances.
+        let max_fees = Amount::from_sat(100000); // Maximum expected mining fees
+        let expected_minimum = original_spendable - max_fees;
+        assert!(
+            balances.spendable >= expected_minimum,
+            "Taproot Maker {} balance should not change significantly. Original spendable: {}, Current spendable: {}",
```
Comment (mojoX911) [inline tests/taproot_multi_maker.rs]: This hugely off. Fees are about 30% of the above max_fee value. We should just assert individual makers' fees.

Comment (mojoX911) [reply tests/taproot_multi_maker.rs]: Same applies to other tests also where this pattern is applied.

Comment (mojoX911) [reply tests/taproot_multi_maker.rs]: We should also add comments mentioning which maker earned which fees.

File: tests/taproot_multi_taker.rs [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+
+    // Verify swap results
+    for (i, taproot_taker) in taproot_takers.into_iter().enumerate() {
+        let taker_balances_after = taproot_taker.get_wallet().get_balances().unwrap();
+        info!("📊 Taproot Taker {} balance after attempting swap:", i);
+        info!(
+            "  Regular: {}, Contract: {}, Spendable: {}",
+            taker_balances_after.regular,
+            taker_balances_after.contract,
+            taker_balances_after.spendable
```
Comment (mojoX911) [inline tests/taproot_multi_taker.rs]: Should have swap balance too in the prints.

**Review by mojoX911 — CHANGES_REQUESTED** [review]

File: tests/taproot_taker_abort3.rs [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+        assert_in_range!(
+            balances.spendable.to_sat(),
+            [14999510, 14999006], // here some fund loss receive by a maker because an outgoing contract was created by it.
+            "Taproot Maker after balance check."
+        );
+
+        let balance_diff = balances
+            .spendable
+            .to_sat()
+            .saturating_sub(original_spendable.to_sat());
```
Comment (mojoX911) [inline tests/taproot_taker_abort3.rs]: We should not staurate sub here. And this applies to all balance diffs. All negative diff values are being suppressed here. That means cases where makers are losing money are not being asserted for the right values.

**Review by mojoX911 — CHANGES_REQUESTED** [review]

File: tests/taproot_swap.rs (hunk 10 above)
Comment (0xEgao) [reply tests/taproot_swap.rs:180]: Here `assert_in_range` is required because in every run most of the time maker's spendable balance, fee gained differs +-2sats. That's why i have used `assert_in_range`. There's no such issues with taker.

Comment (mojoX911) [reply tests/taproot_swap.rs:180]: Do you have a guess why that is? Behaviour here should be similar to takers, as we are just creating taproot txs for the swaps?

Comment (0xEgao) [reply tests/taproot_swap.rs:180]: Once our taproot protocol will get the Taproot Descriptors updated for script path spending, we won't receive this +-2 sats ambiguity i guess.

File: tests/taproot_taker_abort3.rs (hunk 14 above)
Comment (0xEgao) [reply tests/taproot_taker_abort3.rs]: Done

File: tests/taproot_swap.rs
Comment (0xEgao) [reply tests/taproot_swap.rs:106]: Done for the test cases which hit this error

File: tests/taproot_maker_abort2.rs (hunk 6 above)
Comment (0xEgao) [reply tests/taproot_maker_abort2.rs]: Done

File: tests/taproot_maker_abort1.rs (hunk 5 above)
Comment (0xEgao) [reply tests/taproot_maker_abort1.rs]: Done

File: tests/taproot_swap.rs (hunk 3 above)
Comment (0xEgao) [reply tests/taproot_swap.rs]: Done

File: tests/taproot_swap.rs (hunk 4 above)
Comment (0xEgao) [reply tests/taproot_swap.rs]: Done

File: tests/taproot_swap.rs (hunk 2 above)
Comment (0xEgao) [reply tests/taproot_swap.rs]: Done, changed the comment
